mne/morph.py

```diff
@@ -0,0 +1,953 @@
+                              VectorSourceEstimate)
+from .source_space import SourceSpaces
+from .surface import read_surface, read_morph_map, mesh_edges
+from .utils import (logger, verbose, get_subjects_dir, warn as warn_,
+                    has_nibabel, has_dipy)
+
+
+class SourceMorph(object):
+    """Container for source estimate morphs.
+
+    Parameters
+    ----------
+    src : instance of SourceSpaces
+        Information about the respective source space, with src being the
+        corresponding list of SourceSpaces
+    subject_from : str | None
+        Name of the original subject as named in the SUBJECTS_DIR.
+        If None src[0]['subject_his_id]' will be used. The default is None.
+    subject_to : str
+        Name of the subject on which to morph as named in the SUBJECTS_DIR
+        The default is 'fsaverage'.
+    data_from : array | list of two arrays | str | None
+        If morphing a surface source estimate, data_from is the
+        Vertex numbers corresponding to the data.
+        If morphing a volume source estimate data_from is the absolute path
+        where subject_from/mri/brain.mgz is stored
+        If None data_from will be found through subject_from if present. The
+        default is None.
+    data_to : array | list of two arrays | str | None
+        If morphing a surface source estimate , data_to is the
+        Vertex numbers corresponding to the data.
+        If morphing a volume source estimate data_to is the absolute path
+        where subject_to/mri/brain.mgz is stored
+        If None data_to will be found through subject_to if present. The
+        default is None.
+    subjects_dir : str | None
+        Path to SUBJECTS_DIR if it is not set in the environment. The default
+        is None.
+    **options : **kwargs
+        Keyword arguments for the respective morphing function.
+
+    Attributes
+    ----------
+    kind : str | None
+        Kind of source estimate. E.g. 'volume' or 'surface'
+    subjects_dir : str | None
+        The path to the FreeSurfer subjects reconstructions.
+        It corresponds to FreeSurfer environment variable SUBJECTS_DIR.
+    subject_from : str | None
+        Name of the subject from which to morph as named in the SUBJECTS_DIR
+    subject_to : str
+        Name of the subject to which to morph as named in the SUBJECTS_DIR
+    vol_info : dict
+        Information about reference volumes for volumetric morph. Additional
+        information required to transform volume source estimate into
+        Nifti1Image using as_volume
+    src_info : dict
+        Information required to transform volume source estimate into
+        Nifti1Image using as_volume
+    morph_data : Trans | dict
+        If morphing a surface source estimate , morph_data is an instance
+        of Trans to morph the data
+        If morphing a volume source estimate morph_data is a dict storing
+        linear and non-linear mapping information.
```

Comment:
is it a reasonable use case to pass morph_data in the init? for me the purpose of the constructor is actually compute the morph_data.

Reply:
I wrote a structure I would employ:

```
class SourceMorph(object):
    """docstring for SourceMorph"""
    def __init__(self, arg):
        self.arg = arg
        if arg is not None:
            self.morph_data = _compute_morph_data(arg)

    def __call__(self, X):
        return _foobar(X, self.morph_data)


def read_source_morph(fname):
    morph_data = _read_morph_data(fname)
    source_morph = SourceMorph(None)
    source_morph.morph_data = morph_data
```

you see what I mean?

Reply:
see what you mean, but technically it's only adding an additional "layer", that contains what is separate now, right?